Source/WebCore/editing/SpellChecker.cpp

```diff
@@ -210,12 +228,45 @@ void SpellChecker::enqueueRequest(Ref<SpellCheckRequest>&& request)
     m_requestQueue.append(WTFMove(request));
 }
 
-void SpellChecker::didCheck(TextCheckingRequestIdentifier identifier, const Vector<TextCheckingResult>& results)
+static bool containsGrammarResult(TextCheckingResult result, const Vector<TextCheckingResult>& existingResults)
 {
-    ASSERT(m_processingRequest);
-    ASSERT(m_processingRequest->data().identifier() == identifier);
-    if (m_processingRequest->data().identifier() != identifier) {
-        m_requestQueue.clear();
+    bool foundIt = false;
+    for (unsigned i = 0; i < existingResults.size(); i++) {
+        TextCheckingResult existingResult = existingResults[i];
+        TextCheckingType existingResultType = *existingResult.type.toSingleValue();
+        if (existingResultType != TextCheckingType::Grammar || result.range.location != existingResult.range.location || result.range.length != existingResult.range.length || result.details.size() != existingResult.details.size())
+            continue;
+        bool detailsMatch = true;
+        for (unsigned j = 0; detailsMatch && j < result.details.size(); j++) {
+            if (result.details[j].range.location != existingResult.details[j].range.location || result.details[j].range.length != existingResult.details[j].range.length || result.details[j].guesses != existingResult.details[j].guesses)
+                detailsMatch = false;
+        }
+        if (detailsMatch)
+            foundIt = true;
+    }
+    return foundIt;
+}
+
+static bool containsAdditionalGrammarResults(const Vector<TextCheckingResult>& results, const Vector<TextCheckingResult>& existingResults)
+{
+    bool retval = false;
+    for (unsigned i = 0; !retval && i < results.size(); i++) {
+        TextCheckingType resultType = *results[i].type.toSingleValue();
+        if (resultType == TextCheckingType::Grammar && !containsGrammarResult(results[i], existingResults))
+            retval = true;
+    }
+    return retval;
```

Comment:
nit: slight rewrite to avoid the retval on stack
```suggestion
    for (const auto& result : results) {
        TextCheckingType resultType = result.type.toSingleValue();
        if (resultType == TextCheckingType::Grammar && !containsGrammarResult(results[i], existingResults))
            return true;
    }
    return false;
```